// valgrind-tool/lib/ArgumentsProcessor.js
"use strict";
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var path = __importStar(require("path"));
var fs_1 = require("fs");
var ArgumentsProcessor = /** @class */ (function () {
    function ArgumentsProcessor(argv) {
        this.argv = argv;
    }
    ArgumentsProcessor.prototype.getOptions = function () {
        var argv = this.argv;
        var output;
        var commentOnly = false;
        var files = new Array();
        for (var i = 0; i < argv.length; i++) {
            // support --arg="x", --arg=x and --arg='x'
            if (argv[i][0] === '-' && argv[i].indexOf('=') > -1) {
                var slices = argv[i].split('=');
                if (slices[1][0] === '"' || slices[1][0] === "'")
                    slices[1] = slices[1].substring(1, slices[1].length - 1);
                argv.splice.apply(argv, __spreadArrays([i, 1], slices));
                i--;
                continue;
            }
            switch (argv[i]) {
                case '-h':
                case '--help':
                    this.printHelp();
                    return undefined;
                case '--license':
                    fs_1.createReadStream(path.resolve(__dirname, '../LICENSE')).pipe(process.stdout);
                    return undefined;
                case '-o':
                case '--output':
                    output = argv[++i];
                    break;
                case '--comment':
                    commentOnly = true;
                    break;
                default:
                    if (argv[i][0] == '-' || argv[i].substring(0, 2) == '--') {
                        process.stderr.write('valgrind-tool FAILED:\n');
                        process.stderr.write("Invalid option " + argv[i] + ". Try valgrind-tool -h\n");
                        return undefined;
                    }
                    files.push(argv[i]);
            }
        }
        return {
            files: files,
            commentOnly: commentOnly,
            output: output
        };
    };
    ArgumentsProcessor.prototype.printHelp = function () {
        fs_1.createReadStream(path.resolve(__dirname, '../HELP')).pipe(process.stdout);
    };
    return ArgumentsProcessor;
}());
exports["default"] = ArgumentsProcessor;
