// valgrind-tool/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var fs_1 = require("fs");
var path = __importStar(require("path"));
var ValgrindLogParser_1 = __importDefault(require("./ValgrindLogParser"));
var ArgumentsProcessor_1 = __importDefault(require("./ArgumentsProcessor"));
var STDOUT_FILE_DESCRIPTOR = 1;
var STDIN_FILE_DESCRIPTOR = 0;
function run(argv) {
    return __awaiter(this, void 0, void 0, function () {
        function printHeader(fd) {
            if (output !== STDOUT_FILE_DESCRIPTOR) {
                process.stdout.write("Processing " + files[fd] + ":\n");
            }
        }
        var options, output, fd, fds, files, _i, _a, file, fullPath, fd, startTime, _loop_1, _b, fds_1, fd;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    options = new ArgumentsProcessor_1["default"](argv).getOptions();
                    if (!options) {
                        return [2 /*return*/];
                    }
                    if (!!options.output) return [3 /*break*/, 1];
                    output = STDOUT_FILE_DESCRIPTOR;
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, fs_1.promises.open(path.resolve(process.cwd(), options.output), fs_1.constants.W_OK | fs_1.constants.O_CREAT)];
                case 2:
                    fd = (_c.sent()).fd;
                    output = fd;
                    _c.label = 3;
                case 3:
                    fds = new Array();
                    files = new Array();
                    if (!!options.files.length) return [3 /*break*/, 4];
                    fds.push(STDIN_FILE_DESCRIPTOR);
                    return [3 /*break*/, 8];
                case 4:
                    _i = 0, _a = options.files;
                    _c.label = 5;
                case 5:
                    if (!(_i < _a.length)) return [3 /*break*/, 8];
                    file = _a[_i];
                    fullPath = path.resolve(process.cwd(), file);
                    return [4 /*yield*/, fs_1.promises.open(fullPath, fs_1.constants.R_OK)];
                case 6:
                    fd = (_c.sent()).fd;
                    fds.push(fd);
                    files[fd] = fullPath;
                    _c.label = 7;
                case 7:
                    _i++;
                    return [3 /*break*/, 5];
                case 8:
                    startTime = Date.now();
                    _loop_1 = function (fd) {
                        var parser;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    printHeader(fd);
                                    parser = new ValgrindLogParser_1["default"]({
                                        input: fd,
                                        includeHeader: true,
                                        output: output,
                                        commentOnly: options.commentOnly,
                                        eventEmitter: output !== STDOUT_FILE_DESCRIPTOR ? function (event) {
                                            switch (event.type) {
                                                case 'readline':
                                                    process.stdout.cursorTo(0, 0);
                                                    process.stdout.clearLine(0);
                                                    process.stdout.clearScreenDown();
                                                    printHeader(fd);
                                                    process.stdout.cursorTo(0, 1);
                                                    process.stdout.clearLine(0);
                                                    process.stdout.write("Line number " + event.lineNumber + "... (" + event.bytesWritten + " bytes written, " + event.suppressionCount + " suppressions)\n");
                                            }
                                        } : undefined
                                    });
                                    return [4 /*yield*/, parser.parse()];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    };
                    _b = 0, fds_1 = fds;
                    _c.label = 9;
                case 9:
                    if (!(_b < fds_1.length)) return [3 /*break*/, 12];
                    fd = fds_1[_b];
                    return [5 /*yield**/, _loop_1(fd)];
                case 10:
                    _c.sent();
                    _c.label = 11;
                case 11:
                    _b++;
                    return [3 /*break*/, 9];
                case 12:
                    if (output !== STDOUT_FILE_DESCRIPTOR) {
                        process.stdout.write("Done in " + (Date.now() - startTime) + "ms!\n\n");
                    }
                    return [2 /*return*/];
            }
        });
    });
}
run(process.argv.slice(2))["catch"](function (reason) {
    console.error(reason);
});
